algo/src/main/java/org/neo4j/graphalgo/impl/ShortestPathAStar.java

```diff
@@ -87,50 +95,56 @@ private void run(final int goalNodeId, final String propertyKeyLat, final String
-            			updateCosts(source, target, weight + currentNodeCost, heuristic);
-            			if (!closedNodes.contains(target)) {
-            				openNodes.add(target, 0);
-            			}
-            			return true;
-            		});
+                    currentNodeId,
+                    direction,
+                    (source, target, relationshipId, weight) -> {
+                        double neighbourLat = getNodeCoordinate(target, propertyKeyLat);
+                        double neighbourLon = getNodeCoordinate(target, propertyKeyLon);
+                        double heuristic = computeHeuristic(neighbourLat, neighbourLon, goalLat, goalLon);
+                        boolean weightChanged = updateCosts(source, target, weight + currentNodeCost, heuristic);
+                        if (!closedNodes.contains(target)) {
+                            if (weightChanged) {
+                                openNodes.update(target);
+                            } else {
+                                openNodes.add(target, 0);
+                            }
+                        }
+                        return true;
+                    });
             progressLogger.logProgress((double) currentNodeId / (nodeCount - 1));
         }
     }
-    
+
     private double computeHeuristic(final double lat1, final double lon1, final double lat2, final double lon2) {
-    		final int earthRadius = 6371;
-    		final double kmToNM = 0.539957;
-    		final double latDistance = Math.toRadians(lat2 - lat1);
-    		final double lonDistance = Math.toRadians(lon2 - lon1);
-    		final double a = Math.sin(latDistance / 2) * Math.sin(latDistance / 2)
-    	            + Math.cos(Math.toRadians(lat1)) * Math.cos(Math.toRadians(lat2))
-    	            * Math.sin(lonDistance / 2) * Math.sin(lonDistance / 2);
-    	    final double c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
-    	    final double distance = earthRadius * c * kmToNM;
-    	    return distance;
+        final int earthRadius = 6371;
+        final double kmToNM = 0.539957;
+        final double latDistance = Math.toRadians(lat2 - lat1);
+        final double lonDistance = Math.toRadians(lon2 - lon1);
+        final double a = Math.sin(latDistance / 2) * Math.sin(latDistance / 2)
+                + Math.cos(Math.toRadians(lat1)) * Math.cos(Math.toRadians(lat2))
+                * Math.sin(lonDistance / 2) * Math.sin(lonDistance / 2);
+        final double c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
+        final double distance = earthRadius * c * kmToNM;
+        return distance;
     }
-    
+
     private double getNodeCoordinate(final int nodeId, final String coordinateType) {
-    		final long neo4jId = graph.toOriginalNodeId(nodeId);
-    		final Node node = dbService.getNodeById(neo4jId);
-    		return (double) node.getProperty(coordinateType);
+        final long neo4jId = graph.toOriginalNodeId(nodeId);
+        final Node node = dbService.getNodeById(neo4jId);
```

Comment:
@mknblch should we really use non-kernel API here?
And why didn't we load this data initially with the graph?

Reply:
I guess thats my fault. A* was a contribution from the community and since our property-loading-mechanism cannot load multiple properties per node I told the author to use the kernel api instead to get it done. Im going to talk with paul & mark about a change in the Weights-interface. I have some ideas how to adapt our current impl.